core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.persistence.cassandra.query
+
+import java.util.concurrent.ThreadLocalRandom
+
+import akka.annotation.InternalApi
+import akka.persistence.cassandra._
+import akka.stream.stage._
+import akka.stream.{Attributes, Outlet, SourceShape}
+import com.datastax.driver.core.{PreparedStatement, ResultSet, Session}
+
+import scala.collection.immutable.Queue
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContextExecutor
+
+/**
+  * INTERNAL API
+  */
+@InternalApi private[akka] object AllPersistenceIdsStage {
+
+  case object Continue
+
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi private[akka] class AllPersistenceIdsStage(
+    refreshInterval: Option[FiniteDuration],
+    fetchSize: Int,
+    preparedStatement: PreparedStatement,
+    session: Session)
+    extends GraphStage[SourceShape[String]] {
+
+  import AllPersistenceIdsStage._
+
+  val out: Outlet[String] = Outlet("AllPersistenceIds.out")
+
+  val shape: SourceShape[String] = SourceShape(out)
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogicWithLogging(shape) {
+
+      implicit def executionContext: ExecutionContextExecutor =
+        materializer.executionContext
+
+      private var knownPersistenceIds: Set[String] = Set.empty
+      private var maybeResultSet = Option.empty[ResultSet]
+      private var buffer: Queue[String] = Queue.empty[String]
+
+      private val queryCallback: AsyncCallback[ResultSet] =
+        getAsyncCallback[ResultSet] { rs =>
+          maybeResultSet = Some(rs)
+          val available = rs.getAvailableWithoutFetching
+          for (_ <- 1 to available) {
+            val s = rs.one().getString("persistence_id")
+            if (!knownPersistenceIds.contains(s)) {
+              buffer = buffer.enqueue(s)
+              knownPersistenceIds += s
+            }
+          }
+          pushWhileAvailable()
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.getExecutionInfo.getPagingState == null) {
+            complete(out)
+          }
+        }
+
+      private def query(): Unit = {
+        val boundStatement = preparedStatement.bind()
+        boundStatement.setFetchSize(fetchSize)
+        session
+          .executeAsync(boundStatement)
+          .asScala
+          .foreach(queryCallback.invoke)
+      }
+
+      private def pushWhileAvailable(): Unit = {
+        while (buffer.nonEmpty && isAvailable(out)) {
+          val (s, newBuffer) = buffer.dequeue
+          buffer = newBuffer
+          push(out, s)
```

Comment:
It will only ever be possible to push one element here, then there will be a new call to `onPull`.
So this should be named `tryPushOne` instead of `pushWhileAvailable`